[EA] People directory

oetherington opened the pull request:
This PR adds the beta version of the EA people directory ([Figma](https://www.figma.com/file/IRUVQ5D2mKxoW0xZnhXIWv/2024-Q1?node-id=1463%3A12795&mode=dev)).

After deploying, we need to run the following (which I've already done on dev but not staging or prod):
```
./scripts/serverShellCommand.sh "Globals.regenerateUserProfileUpdatedAt()"
./scripts/serverShellCommand.sh "Globals.elasticConfigureIndex('Users')"
./scripts/serverShellCommand.sh "Globals.elasticExportCollection('Users')"
```
The first step takes 20-25 minutes, but the second two are much quicker.

<img width="1511" alt="Screenshot 2024-04-09 at 14 34 55" src="https://github.com/ForumMagnum/ForumMagnum/assets/5075628/c1d18075-fa0c-424e-9a90-0676993a991a">

<img width="733" alt="Screenshot 2024-04-09 at 14 35 15" src="https://github.com/ForumMagnum/ForumMagnum/assets/5075628/ab1e1248-8c3e-4082-95d5-ca594344d2ff">

<img width="659" alt="Screenshot 2024-04-09 at 14 36 06" src="https://github.com/ForumMagnum/ForumMagnum/assets/5075628/c637eae2-311c-4e99-ba50-ea4bc9052555">

<img width="641" alt="Screenshot 2024-04-09 at 14 36 23" src="https://github.com/ForumMagnum/ForumMagnum/assets/5075628/a5a73aaf-12fb-458b-bc14-4b12bef816a6">

<img width="274" alt="Screenshot 2024-04-09 at 14 36 31" src="https://github.com/ForumMagnum/ForumMagnum/assets/5075628/33c1af73-ef46-4c27-a253-9461e3fcd242">

<img width="261" alt="Screenshot 2024-04-09 at 14 36 44" src="https://github.com/ForumMagnum/ForumMagnum/assets/5075628/c0b15e1c-efd6-4f83-b5b0-0fbaebb95585">

<img width="1033" alt="Screenshot 2024-04-09 at 14 36 52" src="https://github.com/ForumMagnum/ForumMagnum/assets/5075628/d32229f0-9da0-47ae-bce5-d756013997d0">


┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1206918564163649) by [Unito](https://www.unito.io)

agnesstenlund commented:
probably didn't state this anywhere in the designs but I was imagining "OR" filter for this dropdown (think it's currently "AND"?)
<img width="582" alt="Screenshot 2024-03-25 at 18 53 07" src="https://github.com/ForumMagnum/ForumMagnum/assets/13235378/a0548a7d-31a6-451c-86eb-9334483229e9">

agnesstenlund commented:
also, this is looking really fecking amazing 🥇

agnesstenlund commented:
these two we can make slimmer by default since they're never above a certain width :) that would leave more space for other columns

<img width="381" alt="Screenshot 2024-03-25 at 19 09 01" src="https://github.com/ForumMagnum/ForumMagnum/assets/13235378/1985b2cd-b0e1-4740-910f-4712fa84df2e">

agnesstenlund commented:
Meta-comment: I'd try to make some of the dropdowns more general components, to make it easier to reuse them in places like these later on (so not making them "people-directory-dropdown" but instead something like "multi-select-dropdown")
![Screenshot 2024-03-26 at 11 03 08](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/376626a5-2e30-4962-b676-e940b054e045)
![Screenshot 2024-03-26 at 11 03 59](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/76a10f10-d0cf-417d-ab6b-62656c7d2181)

![Screenshot 2024-03-26 at 11 13 01](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/30660b59-8adb-45d0-95cc-e069e0d29784)
![Screenshot 2024-03-26 at 11 13 22](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/f5544060-5fda-4347-b461-7317a07c4a51)
![Screenshot 2024-03-26 at 11 13 11](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/bae3c84a-186e-4c83-b6c6-09738926ff8c)

agnesstenlund commented:
> Do you think all the filters should be "OR"?

I do, always "OR" inside of one filter. But "AND" between different filters. So "Location: Stockholm OR Oslo" AND "Org is CEA or 80k" would show people in Stockholm or Oslo who also work at either CEA or 80k

oetherington commented:
@agnesstenlund Another open question: I'm currently excluding the current user from the results, but I'm not sure if this is correct or not

agnesstenlund commented:
> I'm currently excluding the current user from the results, but I'm not sure if this is correct or not

I wouldn't! I think it's good knowing you're showing up + when you show up. So everybody sees the same thing

agnesstenlund commented:
We should also have some smart loading setup, like showing more rows as you scroll, or let people click a "Load more" at the bottom (slightly less nice, I think)

agnesstenlund commented:
Really like the hovers <3

agnesstenlund commented:
maybe this is a local environment thing, but I don't get any results for the locations that show up in the list–I assume we only show locations in the list that people have added? 😮 
![Screenshot 2024-04-04 at 14 07 46](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/b686cac8-0b86-47b7-9497-5821ee286261)
![Screenshot 2024-04-04 at 14 07 32](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/7176997f-b1e7-4693-90e8-07c6070d59b9)

agnesstenlund commented:
just had another look, it's starting to feel _so_ great 🥲 

Some things:

- [ ] For the beta version, I think it'd be really cool to change the "NEW" label to "BETA", so people feel special + more inclined to write feedback :)
![Screenshot 2024-04-05 at 14 31 40](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/cd6e313c-4c0e-4e40-88d2-caf8e56233aa)

- [ ] We should also add a "Give feedback" button to the beta version, either link to Intercom or to a Google form I write
![Screenshot 2024-04-05 at 14 50 54](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/73ca21d0-dd3f-4440-a6fe-cd3d1f2368bd)

- [ ] I like this width a lot for the smaller columns, but realising now they need a bit more padding in between each other
![Screenshot 2024-04-05 at 14 19 04](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/cf7d1bca-43ab-4ca0-bd65-752ef711676e)

- [ ] Would be nice to show a hover text when hovering the message icon–would do "Send message" which will hopefully make it more clear it's a message and not an email
![Screenshot 2024-04-05 at 14 21 39](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/f346e3ab-4215-44f5-ba73-6fd530fee473)

- [ ] Change to "Profile last updated" as per request by Toby. Lets keep "Profile updated" in the table though so the column doesn't get too wide
![Screenshot 2024-04-05 at 14 23 13](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/b24f26b5-2f0e-44a0-a50e-9072344734fc)

- [ ] **Not needed for beta** but there are probably some scenarios where we do want the slim columns to act a bit flexible again, like in this circumstance: 
![Screenshot 2024-04-05 at 14 27 57](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/10299490-d888-409b-9174-36cbc0358b62)

- [ ] **Also not needed for beta** but these don't have a different hover + focused state
![Screenshot 2024-04-05 at 14 29 01](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/8d6b14ca-bf7b-4bba-8da2-e273c683a068)
![Screenshot 2024-04-05 at 14 29 08](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/4b217950-fc25-47a9-9fe8-f8a539873bb5)

agnesstenlund commented:
would be cool to add the number of search results like this for now:
<img width="760" alt="Screenshot 2024-04-08 at 14 56 22" src="https://github.com/ForumMagnum/ForumMagnum/assets/13235378/13f060a2-9f07-4bf6-b914-a110d3bca155">

agnesstenlund commented:
- [ ] oh also the little caret is the wrong color when there is something in the filter:

<img width="214" alt="Screenshot 2024-04-08 at 14 37 31" src="https://github.com/ForumMagnum/ForumMagnum/assets/13235378/88b17dc3-dac8-4132-9bac-00b82fc855a4">

agnesstenlund commented:
- [ ] wouldn't say this is needed for beta, but if it's easy we could implement the empty state
![image](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/a1a84cfd-9afe-4aca-8aa4-fc089b5d4822)

agnesstenlund commented:
- [ ] we also need a different hover state for filled out filters 🙏 

https://github.com/ForumMagnum/ForumMagnum/assets/13235378/305d8a5a-8070-4765-8374-c0b8537e64de

agnesstenlund commented:
the recent changes are so good, thank youu 🙌

agnesstenlund commented:
- [ ] btw "Clear search" in empty state should also clear the filters

agnesstenlund commented:
- [ ] one more tiny thing, I think it'd be slightly nicer if we only show the number of results after you've added any filters or search words
<img width="616" alt="Screenshot 2024-04-09 at 12 40 19" src="https://github.com/ForumMagnum/ForumMagnum/assets/13235378/7734653b-5828-40d1-8b50-454dcdc3da26">

agnesstenlund commented:
- [ ] Before we ship this to beta we should probably hide it on mobile, alternatively show the sidebar thing but then on the actual directory say "This is a beta feature and only works on desktop" instead of showing the directory itself

agnesstenlund commented:
Other than that I think this is ready to ship to beta. We could add the topic interests before, but that needs a bit of a redesign in settings so would probably exclude that for the v1 beta

agnesstenlund commented:
+ link the "share your thoughts" to this google form: https://forms.gle/3RpvUhiwyKFzMSju9 

is there anything else left to do @oetherington?

oetherington commented:
Great! Thanks @agnesstenlund!

I'll make those last few changes then open this up for code review (which might take a while as it's a very big PR). There's nothing else I can think of right now.

agnesstenlund commented:
also the skeleton loading is looking soo nice ✨

Any chance each column could keep their width? Not a big deal if not, it still looks really nice. But that'd avoid a tiny layout shift:

https://github.com/ForumMagnum/ForumMagnum/assets/13235378/f61c5798-561a-4a96-8f7b-05c9f2cf8c9e

oetherington commented:
@agnesstenlund I'll have a think about it, but it's probably not trivial (definitely doable though).

Will-Howard commented on packages/lesswrong/components/peopleDirectory/PeopleDirectoryFilters.tsx:
```diff
@@ -0,0 +1,90 @@
+        }
+      </div>
+      <div className={classes.options}>
+        <PeopleDirectoryFilterDropdown
+          title="Columns"
+          icon="ViewColumns"
+          style="button"
+          className={classes.columnsList}
+        >
+          {columns.filter(({hideable}) => hideable).map((state) => (
+            <PeopleDirectorySelectOption state={state} key={state.value} />
+          ))}
+        </PeopleDirectoryFilterDropdown>
+      </div>
```
I would like a "reset to default" option on this one (cc @agnesstenlund )

Will-Howard reviewed (commented):
Generally looks good, I think there are a few blocking things, I'll do a second pass and note which comments are blocking. Here are few bugs which don't fit anywhere specific:
- When I navigate away and back to the page (i.e. don't do a full reload) it then doesn't autoload more items on scroll, so it's stuck on the first few it loaded
- I got this error when I tried searching "ollie etheri..." (edit: I also got it when filtering by location, the final screenshot is from that case):
![Screenshot 2024-04-15 at 11 48 13](https://github.com/ForumMagnum/ForumMagnum/assets/77623106/5e11de77-4129-488e-b8e8-331d6b93da5a)
![Screenshot 2024-04-15 at 11 48 24](https://github.com/ForumMagnum/ForumMagnum/assets/77623106/a4830835-1a2d-4fff-bf88-7f82ff6ac639)
![Screenshot 2024-04-15 at 12 24 28](https://github.com/ForumMagnum/ForumMagnum/assets/77623106/ed4c0d56-4e3e-48df-be8e-f05a8b67f47a)
- It would be really nice to have scroll restoration when you navigate away and back, but I know this would be quite a big change to do. Failing that, maybe making it easy to open people's profiles in a new tab to nudge people to do that as the default (cc @agnesstenlund)?
- Not sure if I said this already but given the proliferation of acronym org names, I think making this keep the case it was originally in would be good (not sure where the casing is being applied)
![Screenshot 2024-04-15 at 11 54 27](https://github.com/ForumMagnum/ForumMagnum/assets/77623106/d9dbea0b-07c6-4ac5-9361-e7497ceca238)

Blocking things:
- [ ] Navigating away and back to the page causes it to not load more items on scroll
- [ ] The "Cannot read properties of undefined" bug when searching on dev (it may turn out this is a dev only issue)

agnesstenlund commented on packages/lesswrong/components/peopleDirectory/PeopleDirectoryHeading.tsx:
```diff
@@ -0,0 +1,74 @@
+}) => {
+  const {sorting, setSorting} = usePeopleDirectory();
+
+  const onToggleSort = useCallback(() => {
```
Agree, thanks for raising this. We're considering a "most filled out profiles + high karma"-thing as "default sort". Do you think seeing it sorted by "Default" would be enough or would that leave you wondering what that means?

Will-Howard replied:
I think having there be a default sorting that is somewhat opaque is ok (with or without a "Default" label), it's when there is a sorting applied that I think the arrow on the relevant column should show as up or down

oetherington commented on packages/lesswrong/components/peopleDirectory/PeopleDirectoryUserCell.tsx:
```diff
@@ -0,0 +1,60 @@
+    flexGrow: 1,
+  },
+  message: {
+    display: "none", // See `PeopleDirectoryUserCell` for logic to show this
```
Yes, but the relevant css `:hover` selector is in the other file

oetherington commented on packages/lesswrong/components/peopleDirectory/PeopleDirectorySelectOption.tsx:
```diff
@@ -0,0 +1,78 @@
+
+  const {ForumIcon} = Components;
+  return (
+    <div onClick={onClick} className={classNames(classes.root, className)}>
```
At the moment it's just the "check" icon from Heroicons. We could add some padding? This is how it looks with 1px of padding:
<img width="212" alt="Screenshot 2024-04-23 at 12 23 37" src="https://github.com/ForumMagnum/ForumMagnum/assets/5075628/9eeeac66-4449-451f-bd62-427f60a6fd2c">

oetherington commented on packages/lesswrong/components/hooks/useSearchableMultiSelect.ts:
```diff
@@ -0,0 +1,130 @@
+    }
+  }, [captureSearch, facetField]);
+
+  const getWithCache = useLRUCache<string, Promise<string[]>>(fetchSuggestions);
```
Hmm - it's just matching the behaviour of the main searchbox. I think I'm just going to leave it for now and think about it more for when we leave beta.

oetherington commented:
@agnesstenlund I think all comments are now addressed. I also gated it so that only beta users will see the people directory in the main menu - I think this was the plan for releasing?

Still outstanding questions from above:

1. Do we want a "reset to default" option for the columns dropdown?
2. Do we need some indication of which sorting is applied?
3. Does it matter that there's currently no user map? Options:
  a. Revert removing the map from the groups page
  b. Add the map to the people directory be release
  c. Do nothing now, add the map to the people directory soon after release
4. Do we want to adjust the styles of the checkboxes in the filter dropdowns?

agnesstenlund commented:
Thank you!! I think none of these are blocking shipping to beta, but I think most of them should be done soon after.

> 1. Do we want a "reset to default" option for the columns dropdown?

that seems good, although not necessary for beta

> 2. Do we need some indication of which sorting is applied?

I think we do, although I'm fine shipping to beta without

> 3. Does it matter that there's currently no user map?

If either reverting or adding is really quick, I'd do that (there is a small risk people would send feedback about it missing). I do think so few people visit the groups page that it's also not a big deal to have it gone for a couple of weeks.


> 4. Do we want to adjust the styles of the checkboxes in the filter dropdowns?

yes, but IMO also not blocking releasing to beta

something that would be nice to have (to increase discoverability for the beta) is adding this to search:
![image](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/1dc8a1c2-13be-4d5d-9b25-193c1b028f3d)

but also wouldn't consider that blocking shipping to beta ☝️

agnesstenlund commented:
And it's correct only beta users should see it in the main menu 👍

oetherington commented:
@agnesstenlund Great! I just added the "Reset" button for the columns dropdown, reverted the change that removes the community map so it's back on the groups page, and added the link to the main users search page. I think we're ready to go now?

(Also CC @Will-Howard for code approval)

agnesstenlund commented:
something weird happened to the LHS 😮  the people directory only shows for a sec when you refresh

https://github.com/ForumMagnum/ForumMagnum/assets/13235378/46988dfc-2ab2-4d7a-969a-5d344b7a8c4f

agnesstenlund commented:
Looks great to me! 🚢

oetherington commented:
That makes sense that it was fixed after a `yarn install` - we just did a react upgrade (sorry, I should have warned you but totally forgot!)

Will-Howard reviewed (approved):
LGTM. If you're not in a huge rush to deploy it I think showing up and down arrows when sorting by a certain column is the most important thing to improve.

Also is there a reason not to allow sorting by karma? I would assume people would want that, also maybe by "Profile updated" but that seems less important

agnesstenlund commented:
I quite like this UI pattern (that highlights the column title + the direction it's getting sorting)
<img width="700" alt="Screenshot 2024-04-25 at 17 43 00" src="https://github.com/ForumMagnum/ForumMagnum/assets/13235378/ad244a66-aa6f-4e9b-9921-d33823b2e004">

or just showing the arrow in one direction in the primary text color:
<img width="781" alt="Screenshot 2024-04-25 at 17 45 47" src="https://github.com/ForumMagnum/ForumMagnum/assets/13235378/672498e8-decf-45ae-bcd9-5d4b499cb35a">

I also agree it probably make sense to be able to sort by both karma and profile updated, probably looks too opinionated not being able to (I had some idea around how you should only be able to filter by karma, not sort, to not make it too easy to see who's got the most or least, but I think I probably endorse transparency more)
